### ISSUE TITLE
Type safe context extension?

### ISSUE THREAD

**Issue by delaneyj** [hopin-team/twirp-ts#7]
I have some info that changes per request (auth in this example, but also pointers to redis instances) that I'd like to be part of the twirp context.  Maybe `createServerX` should be `createServerX<T extends TwirpContext>` and a then the hooks would have the right type?

**Comment by delaneyj** [comment]
Also could you reexport `TwirpContext` in the generated `x.twirp` package?

**Comment by fenos** [comment]
Great suggestion! Let me have a think how to best implement this.
Surely is something i'd like to see

**Comment by fenos** [comment]
> Also could you reexport TwirpContext in the generated x.twirp package?

Why would you want that to be exported? It is already exported from the `twirp-ts` package

**Comment by fenos** [comment]
This is now implemented in `2.0.0-rc-3` you can do as following:

```ts
interface CustomContext extends TwirpContext {
  customField: string
}

createServer<CustomContext>();
```

it will be then propagated everywhere, (handler, hooks, interceptors)

**Comment by delaneyj** [comment]
> > Also could you reexport TwirpContext in the generated x.twirp package?
> 
> Why would you want that to be exported? It is already exported from the `twirp-ts` package

You actively need it if you are extending the Type.  Just cleans up the imports is all.  

This is great @fenos , you've been so helpful!
